packaging/package-indexer/remote_storage_synchronizer/s3_bucket_synchronizer.py

```diff
@@ -0,0 +1,296 @@
+            remote_root_dir=Path(""),
+            local_root_dir=Path(DEFAULT_ROOT_DIR, bucket),
+        )
+
+    @staticmethod
+    def get_config_keyword() -> str:
+        return "s3"
+
+    @staticmethod
+    def from_config(cfg: dict) -> RemoteStorageSynchronizer:
+        return S3BucketSynchronizer(
+            access_key=cfg["access-key"],
+            secret_key=cfg["secret-key"],
+            endpoint=cfg["endpoint"],
+            bucket=cfg["bucket"],
+        )
+
+    @property
+    def local_files(self) -> List[Path]:
+        dirs_and_files = list(self.local_dir.working_dir.rglob("*"))
+        return list(
+            filter(lambda path: path.is_file() and not path.name.endswith(".package-indexer-md5sum"), dirs_and_files)
+        )
+
+    def __list_existing_objects(self) -> List[Dict[str, Any]]:
+        objects: List[Dict[str, Any]] = []
+        pagination_options: Dict[str, str] = {}
+
+        while True:
+            try:
+                response: Dict[str, Any] = self.__client.list_objects(
+                    Bucket=self.__bucket,
+                    **pagination_options,
+                )
+            except (ClientError, EndpointConnectionError) as e:
+                self._log_error(f"Failed to list objects of bucket: {self.__bucket} => {e}")
+                return []
+
+            try:
+                for obj in response.get("Contents", []):
+                    objects.append(obj)
+                if not response["IsTruncated"]:
+                    break
+                pagination_options = {"Marker": response["Contents"][-1]["Key"]}
+            except KeyError:
+                self._log_error(
+                    f"Failed to list objects of bucket: {self.__bucket}/ => Unexpected response: {response}"
+                )
+                return []
+
+        return objects
+
+    @property
+    def remote_files(self) -> List[Dict[str, Any]]:
+        if self.__remote_files_cache is not None:
+            return self.__remote_files_cache
+
+        self.__remote_files_cache = self.__list_existing_objects()
+        return self.__remote_files_cache
+
+    def __get_remote_md5sum_file_path(self, relative_file_path: str) -> Path:
+        path = Path(self.local_dir.root_dir, "package-indexer-md5sums", relative_file_path).resolve()
+        return Path(path.parent, path.name + ".package-indexer-md5sum")
+
+    def __download_file(self, relative_file_path: str) -> None:
+        download_path = Path(self.local_dir.root_dir, relative_file_path).resolve()
+
+        self._log_info(
+            "Downloading file.",
+            remote_path=relative_file_path,
+            local_path=str(download_path),
+        )
+
+        download_path.parent.mkdir(parents=True, exist_ok=True)
+        with download_path.open("wb") as downloaded_object:
+            self.__client.download_fileobj(self.__bucket, relative_file_path, downloaded_object)
+
+        md5sum = md5(download_path.read_bytes()).digest()
```

Comment:
Since we are the ones calculating the remote hashes, can't we use something more modern and trusted, like sha256?
I know it's not important in this use case because we have endless layers of integrity checks (and PGP signature as well), but it would just look better if we didn't see md5 in this unit.

Reply:
Sure, done!